### PR TITLE
Added filter functionality

### PR DESCRIPTION
**Pull request by AshutoshM29**
I have finally added filter functionality for the e-com!! Please give feedback on code and UI too!! :)

Files to check:
[Sidebar](src/components/Sidebar/Sidebar.jsx)
[Product Context](src/context/product-context.jsx)

Files to avoid: 
BIN files
[Index](public/index.html)
[App-CSS](src/App.css)

### CODE REVIEW

File: src/components/Sidebar/Sidebar.jsx [hunk 1]
```diff
@@ -1,80 +1,146 @@
...
+                        <label className="input-desc">2 Stars & above</label>
+                    </div>
+                    <div className='checkbox'>
+                        <input
+                            className="input-checkbox"
+                            type="checkbox"
+                            checked={state.rating1}
+                            onClick={()=>rating1Handler(dispatch)}
+                        />
+                        <label className="input-desc">1 Stars & above</label>
```
Comment (Ankur9669) [inline src/components/Sidebar/Sidebar.jsx]: Use htmlFor for labels

File: src/context/product-context.jsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { createContext, useReducer, useContext } from "react";
+import { useProducts } from "../hooks/hooks";
+import { filterProducts } from "../hooks/Filters/category-filter";
+import { discountedItems } from "../hooks/Filters/discount-price-filter";
+import { pricedProducts } from "../hooks/Filters/price-filter";
+import { reducerFunction } from "../hooks/Filters/reducer-function";
+import { SortedProducts } from "../hooks/Filters/sort-price-filter";
+import { starredProducts } from "../hooks/Filters/rating-filter";
+
```
Comment (Ankur9669) [inline src/context/product-context.jsx:9]: You can import all files in single file

File: src/context/product-context.jsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+export function ProductProvider({ children }) {
+    const [state, dispatch] = useReducer(reducerFunction, reducerValues)
+    const { products } = useProducts()
+    const filteredProducts = filterProducts(products, state.sortBy)
+    const starProducts = starredProducts(filteredProducts, state)
+    const priceProducts = pricedProducts(starProducts, state.priceSelector)
+    const discountItems = discountedItems(priceProducts, state.discountValue)
+    const sortPriceHighLow = SortedProducts(discountItems, state.priceHighLow)
+
+    return <ProudctContext.Provider value={{ state, dispatch, sortPriceHighLow }}>
```
Comment (Ankur9669) [inline src/context/product-context.jsx:30]: fix: product spelling is wrong

File: src/hooks/Filters/filter-complete.jsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+}
+    
+function rating2Handler(dispatch) {
+dispatch({ type: "rating2", payload: "rating2" })
+}
+    
+function rating1Handler(dispatch) {
+dispatch({ type: "rating1", payload: "rating1" })
+}
+
```
Comment (Ankur9669) [inline src/hooks/Filters/filter-complete.jsx:24]:
```
function rating4Handler(dispatch) {
dispatch({ type: "rating4", payload: "rating4" })
}

function rating3Handler(dispatch) {
dispatch({ type: "rating3", payload: "rating3" })
}

function rating2Handler(dispatch) {
dispatch({ type: "rating2", payload: "rating2" })
}

function rating1Handler(dispatch) {
dispatch({ type: "rating1", payload: "rating1" })
}
```

Suggestion: This is also fine, but these functions could be written into single function called ratinghandler and in that function you can pass the "type" and "payload" as parameter as use it

File: src/hooks/Filters/rating-filter.jsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        return SortedProducts.filter (items => {
+            return (items.rating >= 3);
+        })
+    }
+    if (stars4) {
+        return SortedProducts.filter (({ rating }) => {
+            return (rating >= 4);
+        })
+    }
+
```
Comment (Ankur9669) [inline src/hooks/Filters/rating-filter.jsx:23]:
```
    if (stars1) {
        return SortedProducts.filter (items => {
            return (items.rating >= 1);
        })
    }
    if (stars2) {
        return SortedProducts.filter (items => {
            return (items.rating >= 2);
        })
    }
    if (stars3) {
        return SortedProducts.filter (items => {
            return (items.rating >= 3);
        })
    }
    if (stars4) {
        return SortedProducts.filter (({ rating }) => {
            return (rating >= 4);
        })
    }
```

These can also be done in one line where you will pass the rating in the function

But for the time being you can keep it this way only

**Review by Ankur9669 — APPROVED** [review]
The code is looking good, but their is some redundancy which could be improved, overall nice work.

I could not find link for UI, so could not review the UI

**Review by Sahil25061999 — APPROVED** [review]
same suggestions as Ankur, I think for input elements like checkbox and slider you need to provide onChange attribute instead of onClick.

**Pull request merged**